src/contracts/logic/math.rs

```
@@ -0,0 +1,444 @@
+use decimal::*;
+
+use crate::math::liquidity::Liquidity;
+use crate::math::sqrt_price::sqrt_price::{calculate_sqrt_price, SqrtPrice};
+use crate::math::token_amount::TokenAmount;
+use crate::math::MAX_TICK;
+
+const PRICE_DENOMINATOR: u128 = 1_000_000_000_000_000_000_000_000;
+const LIQUIDITY_DENOMINATOR: u128 = 1_000_000;
+
+// taken from protocol
+const TICK_LIMIT: i32 = 44_364;
+
+pub fn get_liquidity(
+    x: TokenAmount,
+    y: TokenAmount,
+    mut lower_tick: i32,
+    mut upper_tick: i32,
+    current_sqrt_price: SqrtPrice,
+    rounding_up: bool,
+    tick_spacing: u16,
+) -> (TokenAmount, TokenAmount, Liquidity) {
+    if lower_tick < -MAX_TICK {
```

It will be better to throw an Error and return `Result`